File: frontend/src/api/costCoverage.ts
import { fetchJson } from "../lib/api/client";

export type CostCoverageStatus =
  | "needs_input"
  | "research_ready"
  | "researched"
  | "estimated"
  | "evidenced"
  | "complete"
  | "not_applicable";

export type ResearchSource = {
  title: string;
  url: string;
};

export type ResearchOption = {
  name: string;
  unit_rate: number | null;
  unit: string;
  estimated_total: number | null;
  notes: string;
  source_url: string;
  details?: Record<string, string | number | boolean | null>;
};

export type CostCoverageRequirement = {
  code: string;
  category: string;
  title: string;
  summary: string;
  collection_mode: "automatic" | "researchable" | "traveler";
  evidence_kind: string;
  required_inputs: string[];
  output_fields: string[];
  research_prompt?: string | null;
  policy_reference?: string | null;
  status: CostCoverageStatus;
  inputs: Record<string, string>;
  missing_inputs: string[];
  estimate_amount: number | null;
  currency: string;
  note: string;
  source_url: string;
  research: {
    status: string;
    summary: string;
    options: ResearchOption[];
    sources: ResearchSource[];
    researched_at: string;
    model: string | null;
  } | null;
  selected_option: ResearchOption | null;
  updated_at: string | null;
};

export type CostCoverageResponse = {
  trip_id: string;
  contract_version: string;
  source_status: string;
  summary: {
    requirement_count: number;
    resolved_count: number;
    research_offer_count: number;
    ready_for_handoff: boolean;
    headline: string;
  };
  requirements: CostCoverageRequirement[];
  research_notice?: {
    status: string;
    missing_inputs: string[];
    message: string;
  } | null;
};

export function fetchCostCoverage(tripId: string): Promise<CostCoverageResponse> {
  return fetchJson<CostCoverageResponse>({
    path: `/api/workspace/${tripId}/cost-coverage`,
    credentials: "include",
  });
}

export function researchCostCoverage(
  tripId: string,
  requirementCode: string,
  inputs: Record<string, string>
): Promise<CostCoverageResponse> {
  return fetchJson<CostCoverageResponse>({
    path: `/api/workspace/${tripId}/cost-coverage/${requirementCode}/research`,
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ inputs }),
  });
}

export function updateCostCoverage(
  tripId: string,
  requirementCode: string,
  payload: {
    status?: CostCoverageStatus;
    estimate_amount?: number;
    currency?: string;
    note?: string;
    source_url?: string;
    inputs?: Record<string, string>;
    selected_option?: ResearchOption;
  }
): Promise<CostCoverageResponse> {
  return fetchJson<CostCoverageResponse>({
    path: `/api/workspace/${tripId}/cost-coverage/${requirementCode}`,
    method: "PATCH",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

File: trip_planner/app/schemas/cost_coverage.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

CoverageStatus = Literal[
    "needs_input",
    "research_ready",
    "researched",
    "estimated",
    "evidenced",
    "complete",
    "not_applicable",
]


class CostCoverageUpdateRequest(BaseModel):
    status: CoverageStatus | None = None
    estimate_amount: float | None = Field(default=None, ge=0)
    currency: str = Field(default="USD", min_length=3, max_length=3)
    note: str | None = None
    source_url: str | None = None
    inputs: dict[str, str] = Field(default_factory=dict)
    selected_option: dict[str, Any] | None = None


class CostCoverageResearchRequest(BaseModel):
    inputs: dict[str, str] = Field(default_factory=dict)


class CostCoverageResponse(BaseModel):
    trip_id: str
    contract_version: str
    source_status: str
    summary: dict[str, Any]
    requirements: list[dict[str, Any]]
    research_notice: dict[str, Any] | None = None
